### PR TITLE
fix nest openmp performance bug in thnn_conv2d

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52577 fix nest openmp performance bug in thnn_conv2d**

Differential Revision: [D27063800](https://our.internmc.facebook.com/intern/diff/D27063800)

### CODE REVIEW

**Comment by mingfeima** [comment]
`at::thnn_conv2d` has very low performance on multi-core CPU when `batch_size=1`.

`at::thnn_conv2d` used the gemm-based algorithm (im2col + gemm), which will have a nested openmp loop: 
- outer loop will parallel on dimension of N (batch_size)
- inner loop will do im2col + gemm per each image, the inner loop is forced to go sequential

This is problematic for **single batch size inference** on a **multi-core** run where available HW resources can not be full utilized. This patch will improve `at::thnn_conv2d` and `at::slow_conv_transpose2d` with #51185, also the 3d counterpart.

For `at::thnn_conv2d`, single batch inference improvements:
- avx512 CPU (20 cores): ~11x
- avx2 CPU (12 cores): ~9x

[TODO] The threading algorithm in `at::parallel_for` introduces load imbalance on some occasions, will port **balance211** algorithm from mkldnn to pytorch in following patches.

**Comment by mingfeima** [comment]
- at::thnn_conv2d performance improvement on **avx512 CPU** (single batch size, 20 cores):

input size | output channels | kernel | stride | padding | groups | before (nchw) | after (nchw) | speedup
-- | -- | -- | -- | -- | -- | -- | -- | --
[1,3,224,224] | 64 | 7 | 2 | 3 | 1 | 5.416 | 0.554 | 9.77
[1,64,56,56] | 128 | 1 | 1 | 0 | 1 | 0.536 | 0.080 | 6.73
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 2.784 | 1.244 | 2.24
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 1.741 | 0.154 | 11.30
[1,64,56,56] | 256 | 1 | 1 | 0 | 1 | 1.104 | 0.125 | 8.82
[1,256,56,56] | 128 | 1 | 1 | 0 | 1 | 1.747 | 0.163 | 10.69
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 2.786 | 1.239 | 2.25
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 1.748 | 0.154 | 11.32
[1,256,56,56] | 128 | 1 | 1 | 0 | 1 | 1.736 | 0.164 | 10.61
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 2.792 | 1.241 | 2.25
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 1.748 | 0.155 | 11.27
[1,256,56,56] | 256 | 1 | 1 | 0 | 1 | 3.283 | 0.247 | 13.27
[1,256,56,56] | 256 | 3 | 2 | 1 | 32 | 6.323 | 1.314 | 4.81
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.608 | 0.140 | 11.50
[1,256,56,56] | 512 | 1 | 2 | 0 | 1 | 2.075 | 0.189 | 10.98
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 1.677 | 0.165 | 10.18
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 2.228 | 1.117 | 1.99
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.607 | 0.140 | 11.50
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 1.675 | 0.165 | 10.16
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 2.228 | 1.110 | 2.01
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.607 | 0.140 | 11.50
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 1.676 | 0.165 | 10.18
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 2.227 | 1.110 | 2.01
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 1.608 | 0.140 | 11.45
[1,512,28,28] | 512 | 1 | 1 | 0 | 1 | 3.081 | 0.253 | 12.18
[1,512,28,28] | 512 | 3 | 2 | 1 | 32 | 3.575 | 1.160 | 3.08
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.775 | 0.146 | 12.19
[1,512,28,28] | 1024 | 1 | 2 | 0 | 1 | 1.947 | 0.176 | 11.08
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.870 | 0.174 | 10.75
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.280 | 1.141 | 2.00
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.776 | 0.145 | 12.21
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.867 | 0.175 | 10.68
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.270 | 1.131 | 2.01
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.773 | 0.146 | 12.13
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.868 | 0.174 | 10.76
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.273 | 1.144 | 1.99
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.775 | 0.146 | 12.15
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.869 | 0.174 | 10.73
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.284 | 1.153 | 1.98
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.776 | 0.146 | 12.17
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.869 | 0.175 | 10.66
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.264 | 1.150 | 1.97
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.776 | 0.146 | 12.18
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.868 | 0.174 | 10.76
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.265 | 1.160 | 1.95
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.775 | 0.146 | 12.16
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.867 | 0.173 | 10.79
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.283 | 1.162 | 1.96
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.772 | 0.146 | 12.15
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.865 | 0.175 | 10.66
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.251 | 1.134 | 1.98
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.772 | 0.146 | 12.11
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.869 | 0.173 | 10.79
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.282 | 1.155 | 1.97
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.771 | 0.146 | 12.14
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.864 | 0.173 | 10.74
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.262 | 1.138 | 1.99
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.770 | 0.146 | 12.09
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.870 | 0.173 | 10.79
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.265 | 1.143 | 1.98
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.772 | 0.146 | 12.16
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.868 | 0.174 | 10.72
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.251 | 1.157 | 1.94
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.772 | 0.146 | 12.14
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.870 | 0.173 | 10.79
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.261 | 1.162 | 1.95
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.774 | 0.147 | 12.06
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.868 | 0.174 | 10.72
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.254 | 1.142 | 1.97
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.776 | 0.146 | 12.18
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.868 | 0.173 | 10.78
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.262 | 1.146 | 1.97
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.771 | 0.146 | 12.16
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.869 | 0.174 | 10.77
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.266 | 1.156 | 1.96
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.771 | 0.146 | 12.13
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.868 | 0.174 | 10.76
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.254 | 1.159 | 1.94
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.772 | 0.146 | 12.14
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.867 | 0.174 | 10.72
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.264 | 1.166 | 1.94
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.772 | 0.146 | 12.13
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.868 | 0.173 | 10.79
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.255 | 1.149 | 1.96
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.775 | 0.146 | 12.15
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.870 | 0.174 | 10.74
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.262 | 1.148 | 1.97
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.773 | 0.146 | 12.16
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.867 | 0.173 | 10.78
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.258 | 1.152 | 1.96
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.774 | 0.146 | 12.16
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 1.870 | 0.174 | 10.75
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.253 | 1.152 | 1.96
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 1.772 | 0.146 | 12.11
[1,1024,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.468 | 0.268 | 12.92
[1,1024,14,14] | 1024 | 3 | 2 | 1 | 32 | 2.775 | 1.311 | 2.12
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 2.359 | 0.189 | 12.48
[1,1024,14,14] | 2048 | 1 | 2 | 0 | 1 | 2.433 | 0.205 | 11.85
[1,2048,7,7] | 1024 | 1 | 1 | 0 | 1 | 2.440 | 0.195 | 12.51
[1,1024,7,7] | 1024 | 3 | 1 | 1 | 32 | 2.430 | 1.297 | 1.87
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 2.361 | 0.189 | 12.48
[1,2048,7,7] | 1024 | 1 | 1 | 0 | 1 | 2.440 | 0.195 | 12.52
[1,1024,7,7] | 1024 | 3 | 1 | 1 | 32 | 2.430 | 1.298 | 1.87
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 2.366 | 0.188 | 12.61


- at::thnn_conv2d performance improvement on **avx2 CPU** (single batch size, 12 cores):

input size | output channels | kernel | stride | padding | groups | before (nchw) | after (nchw) | speedup
-- | -- | -- | -- | -- | -- | -- | -- | --
[1,3,224,224] | 64 | 7 | 2 | 3 | 1 | 8.555 | 1.694 | 5.05
[1,64,56,56] | 128 | 1 | 1 | 0 | 1 | 1.059 | 0.158 | 6.71
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 3.674 | 2.270 | 1.62
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 3.270 | 0.364 | 8.97
[1,64,56,56] | 256 | 1 | 1 | 0 | 1 | 2.198 | 0.228 | 9.64
[1,256,56,56] | 128 | 1 | 1 | 0 | 1 | 3.241 | 0.320 | 10.12
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 3.607 | 1.881 | 1.92
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 3.507 | 0.337 | 10.41
[1,256,56,56] | 128 | 1 | 1 | 0 | 1 | 3.263 | 0.308 | 10.61
[1,128,56,56] | 128 | 3 | 1 | 1 | 32 | 3.640 | 1.880 | 1.94
[1,128,56,56] | 256 | 1 | 1 | 0 | 1 | 3.229 | 0.355 | 9.09
[1,256,56,56] | 256 | 1 | 1 | 0 | 1 | 5.928 | 0.580 | 10.22
[1,256,56,56] | 256 | 3 | 2 | 1 | 32 | 6.200 | 1.672 | 3.71
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 2.839 | 0.314 | 9.05
[1,256,56,56] | 512 | 1 | 2 | 0 | 1 | 3.141 | 0.410 | 7.66
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 2.956 | 0.377 | 7.84
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 3.130 | 1.628 | 1.92
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 2.810 | 0.315 | 8.91
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 2.992 | 0.360 | 8.32
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 3.070 | 1.612 | 1.90
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 2.807 | 0.314 | 8.95
[1,512,28,28] | 256 | 1 | 1 | 0 | 1 | 2.910 | 0.374 | 7.77
[1,256,28,28] | 256 | 3 | 1 | 1 | 32 | 3.074 | 1.616 | 1.90
[1,256,28,28] | 512 | 1 | 1 | 0 | 1 | 2.798 | 0.314 | 8.92
[1,512,28,28] | 512 | 1 | 1 | 0 | 1 | 5.450 | 0.590 | 9.24
[1,512,28,28] | 512 | 3 | 2 | 1 | 32 | 4.073 | 1.661 | 2.45
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.040 | 0.346 | 8.79
[1,512,28,28] | 1024 | 1 | 2 | 0 | 1 | 3.218 | 0.390 | 8.24
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.082 | 0.396 | 7.78
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.736 | 1.694 | 1.62
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.095 | 0.340 | 9.11
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.162 | 0.385 | 8.21
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.715 | 1.690 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.055 | 0.347 | 8.81
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.061 | 0.399 | 7.68
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.653 | 1.680 | 1.58
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.542 | 0.343 | 10.31
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.473 | 0.384 | 9.05
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.755 | 1.673 | 1.65
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.028 | 0.340 | 8.90
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.122 | 0.386 | 8.10
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.713 | 1.687 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.052 | 0.343 | 8.90
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.108 | 0.380 | 8.18
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.698 | 1.689 | 1.60
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.114 | 0.346 | 8.99
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.105 | 0.391 | 7.93
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.732 | 1.696 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.036 | 0.340 | 8.93
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.162 | 0.385 | 8.22
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.711 | 1.687 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.080 | 0.342 | 9.00
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.126 | 0.380 | 8.23
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.777 | 1.669 | 1.66
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.047 | 0.342 | 8.92
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.159 | 0.391 | 8.08
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.706 | 1.676 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.022 | 0.342 | 8.84
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.088 | 0.385 | 8.03
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 3.300 | 1.680 | 1.96
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.265 | 0.353 | 9.25
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.132 | 0.391 | 8.01
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.646 | 1.670 | 1.58
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.050 | 0.340 | 8.96
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.092 | 0.386 | 8.02
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.711 | 1.674 | 1.62
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.074 | 0.344 | 8.93
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.084 | 0.383 | 8.06
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.654 | 1.673 | 1.59
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.082 | 0.343 | 8.98
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.090 | 0.383 | 8.06
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.681 | 1.685 | 1.59
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.089 | 0.342 | 9.02
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.129 | 0.384 | 8.14
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.735 | 1.665 | 1.64
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.086 | 0.340 | 9.07
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.087 | 0.381 | 8.09
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.671 | 1.693 | 1.58
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.112 | 0.346 | 9.00
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.103 | 0.383 | 8.09
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.722 | 1.664 | 1.64
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.028 | 0.344 | 8.80
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.111 | 0.393 | 7.91
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.651 | 1.723 | 1.54
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.069 | 0.341 | 9.00
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.119 | 0.382 | 8.17
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.683 | 1.665 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.033 | 0.343 | 8.84
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.072 | 0.391 | 7.87
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.703 | 1.681 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.003 | 0.340 | 8.83
[1,1024,14,14] | 512 | 1 | 1 | 0 | 1 | 3.077 | 0.385 | 8.00
[1,512,14,14] | 512 | 3 | 1 | 1 | 32 | 2.684 | 1.668 | 1.61
[1,512,14,14] | 1024 | 1 | 1 | 0 | 1 | 3.032 | 0.344 | 8.80
[1,1024,14,14] | 1024 | 1 | 1 | 0 | 1 | 5.931 | 0.642 | 9.23
[1,1024,14,14] | 1024 | 3 | 2 | 1 | 32 | 3.355 | 2.038 | 1.65
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 3.793 | 0.418 | 9.08
[1,1024,14,14] | 2048 | 1 | 2 | 0 | 1 | 3.863 | 0.532 | 7.26
[1,2048,7,7] | 1024 | 1 | 1 | 0 | 1 | 3.840 | 0.434 | 8.85
[1,1024,7,7] | 1024 | 3 | 1 | 1 | 32 | 2.768 | 1.850 | 1.50
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 3.858 | 0.376 | 10.25
[1,2048,7,7] | 1024 | 1 | 1 | 0 | 1 | 3.912 | 0.388 | 10.07
[1,1024,7,7] | 1024 | 3 | 1 | 1 | 32 | 2.772 | 1.774 | 1.56
[1,1024,7,7] | 2048 | 1 | 1 | 0 | 1 | 3.863 | 0.379 | 10.20

File: aten/src/ATen/ParallelOpenMP.h [hunk 1]
```diff
@@ -22,6 +22,10 @@ inline void parallel_for(
   if (begin >= end) {
     return;
   }
+  if (end - begin == 1) {
```
Comment (jgong5) [inline aten/src/ATen/ParallelOpenMP.h:25]: Shall we check `end - being <= grain_size` instead?

**Comment by jgong5** [comment]
Copy @VitalyFedyunin

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**